src/maps.c

```diff
@@ -273,6 +274,56 @@ unsigned int decl_map_count(enum decl_flavor flavor)
 	}
 }
 
+no_sanitize_unsigned_integer_
+void mark_userspace_class(const char *class_name)
+{
+	struct bool_hash_elem *userspace_class;
+
+	HASH_FIND(hh_userspace_class, userspace_class_map, class_name, strlen(class_name), userspace_class);
+
+	if (!userspace_class) {
+		userspace_class = malloc(sizeof(struct bool_hash_elem));
+		userspace_class->key = strdup(class_name);
+		userspace_class->val = 1;
+		HASH_ADD_KEYPTR(hh_userspace_class, userspace_class_map, userspace_class->key,
+				strlen(userspace_class->key), userspace_class);
+	} else {
+		userspace_class->val = 1;
+	}
+}
+
+no_sanitize_unsigned_integer_
+int is_userspace_class(const char *class_name, const struct string_list *permissions)
```

Comment:
I think this is how it works, but just to make sure I'm understanding correctly:

This function returns 1 if we've positively marked something as a userspace class, correct?  So if we encounter some class `foo` and don't know about it, we return false here, and then don't lint on it needing a requirement?

If so, I think that's the desired behavior.  I'd like to see that clarified in a comment.

I'm particularly concerned about the case where we didn't find the security_classes file for some reason.  I think in that case, we should warn (as the code does) and then skip the userspace checks.  It looks like everything is doing that, but I wanted to double check to make sure, and make sure that gets in a comment for posterity.  Maybe a comment by the warning would be helpful as well.

Reply:
Oh, actually, that's not the behavior.  Don't we hit the "Class %s is listed in require block but is not a userspace class" message if we fail to load the security classes?  I feel like if we didn't load any security classes, we should just skip checking the class requires, rather than reporting on every single class require.

I haven't thought this through thoroughly, but maybe just having this function always return 0 if there are no userspace classes would be sufficient?

Reply:
The function `is_userspace_class()` returns `1` if `class_name` is a known userspace class or it is the `system` class with a permission not to be known as a kernel permission, else `0`.
Added a global flag `userspace_class_support` whether `security_classes` was successfully parsed to skip checks depending on userspace class support, since there is an instance where a return value from `is_userspace_class()` of `0` skips a warning, and one for the return value of `1`.